internal/core/src/segcore/FieldIndexing.cpp

```diff
@@ -78,20 +83,78 @@ VectorFieldIndexing::GetDataFromIndex(const int64_t* seg_offsets,
     ids_ds->SetDim(1);
     ids_ds->SetIds(seg_offsets);
     ids_ds->SetIsOwner(false);
+    if (field_meta_.get_data_type() == DataType::VECTOR_SPARSE_FLOAT) {
+        auto vector = index_->GetSparseVector(ids_ds);
+        SparseRowsToProto(
+            [vec_ptr = vector.get()](size_t i) { return vec_ptr + i; },
+            count,
+            reinterpret_cast<milvus::proto::schema::SparseFloatArray*>(output));
+    } else {
+        auto vector = index_->GetVector(ids_ds);
+        std::memcpy(output, vector.data(), count * element_size);
+    }
+}
 
-    auto vector = index_->GetVector(ids_ds);
+void
+VectorFieldIndexing::AppendSegmentIndexSparse(int64_t reserved_offset,
+                                              int64_t size,
+                                              int64_t new_data_dim,
+                                              const VectorBase* field_raw_data,
+                                              const void* data_source) {
+    auto conf = get_build_params();
+    auto source = dynamic_cast<const ConcurrentVector<SparseFloatVector>*>(
+        field_raw_data);
+    AssertInfo(source,
+               "field_raw_data can't cast to "
+               "ConcurrentVector<SparseFloatVector> type");
+    AssertInfo(size > 0, "append 0 sparse rows to index is not allowed");
+    if (!built_) {
+        AssertInfo(!sync_with_index_, "index marked synced before built");
+        idx_t total_rows = reserved_offset + size;
+        idx_t chunk_id = 0;
+        auto dim = source->Dim();
+
+        while (total_rows > 0) {
+            auto mat = static_cast<const knowhere::sparse::SparseRow<float>*>(
+                source->get_chunk_data(chunk_id));
+            auto rows = std::min(source->get_size_per_chunk(), total_rows);
+            auto dataset = knowhere::GenDataSet(rows, dim, mat);
+            dataset->SetIsSparse(true);
+            try {
+                if (chunk_id == 0) {
+                    index_->BuildWithDataset(dataset, conf);
+                } else {
+                    index_->AddWithDataset(dataset, conf);
+                }
+            } catch (SegcoreError& error) {
+                LOG_ERROR("growing sparse index build error: {}", error.what());
```

Comment:
what happen if build or add error?

Will we rebuild everthing from chunk 0? will there be duplicated data?

Reply:
this is related to both dense and sparse, I am updating it in https://github.com/milvus-io/milvus/pull/31170.